File: lib/src/main/java/com/sd/lib/http/utils/HttpDataHolder.java
package com.sd.lib.http.utils;

import java.util.LinkedHashMap;
import java.util.Map;

public class HttpDataHolder<K, V>
{
    private final Map<K, V> mMap = new LinkedHashMap<>();

    public HttpDataHolder<K, V> put(K key, V value)
    {
        if (value == null)
        {
            mMap.remove(key);
        } else
        {
            mMap.put(key, value);
        }
        return this;
    }

    public HttpDataHolder<K, V> put(Map<K, V> map)
    {
        if (map != null)
        {
            for (Map.Entry<K, V> item : map.entrySet())
            {
                put(item.getKey(), item.getValue());
            }
        }
        return this;
    }

    public HttpDataHolder<K, V> put(HttpDataHolder<K, V> data)
    {
        if (data != null)
            put(data.toMap());
        return this;
    }

    public V get(Object key)
    {
        return mMap.get(key);
    }

    public int size()
    {
        return mMap.size();
    }

    public HttpDataHolder<K, V> clear()
    {
        mMap.clear();
        return this;
    }

    public Map<K, V> toMap()
    {
        return new LinkedHashMap<>(mMap);
    }
}

File: lib/src/main/java/com/sd/lib/http/utils/HttpDataHolder.kt
package com.sd.lib.http.utils

import java.util.*

class HttpDataHolder<K, V> {
    private val _map = mutableMapOf<K, V>()

    fun put(key: K, value: V?): HttpDataHolder<K, V> {
        if (value == null) {
            _map.remove(key)
        } else {
            _map[key] = value
        }
        return this
    }

    fun put(map: Map<K, V>?): HttpDataHolder<K, V> {
        if (map != null) {
            for ((key, value) in map) {
                put(key, value)
            }
        }
        return this
    }

    fun put(data: HttpDataHolder<K, V>?): HttpDataHolder<K, V> {
        data?.let {
            put(it.toMap())
        }
        return this
    }

    operator fun get(key: Any): V? {
        return _map[key]
    }

    fun size(): Int {
        return _map.size
    }

    fun clear(): HttpDataHolder<K, V> {
        _map.clear()
        return this
    }

    fun toMap(): Map<K, V> {
        return LinkedHashMap(_map)
    }
}